### PR TITLE
Speed up pyhit value retrieval for faster Test Harness startup

### LINKED ISSUE #26068: Speed up the Test Harness startup time
## Reason

The integration of WASP-HIT into pyhit and the Test Harness in https://github.com/idaholab/moose/pull/25432 slowed down the startup time of testing.

## Design

Speed up hot spot in HIT Node data access used by pyhit.

## Impact

This will reduce the startup time of testing.

### PR DESCRIPTION
**Pull request by brandonlangley**
closes #26068

### CODE REVIEW

**Comment by brandonlangley** [comment]
@permcody @loganharbour -

The retrieval of values from the HIT parse tree with C++ code is templated based on the expected value type, e.g.,

```cpp
std::vector<std::string> strs = root->param<std::vector<std::string>>(path)

double num = node->param<double>(name))
```

However, pyhit relies extensively on the `Field::kind()` method to get parameter values from the parse tree, e.g.,

`framework/contrib/hit/hit.pyx:`
```python
k = nn.kind()
if k == FieldKind.Int:
    return int(f.intVal())
elif k == FieldKind.Float:
    return float(f.floatVal())
elif k == FieldKind.Bool:
    return bool(f.boolVal())
return f.strVal().decode('utf-8')
```

This updates `Field::kind()` to use the underlying WASP token type instead of deducing kind from a string value.

Since it's so heavily used by Python for types when getting values, this speeds up my Mac's Test Harness startup:

 - from ~24 seconds down to ~3 seconds for `moose/test`

 - from ~46 seconds down to ~5 seconds for `moose/modules`

Could I ask one of you to review this to merge?

**Review by dschwen — APPROVED** [review]
Looks good to me

**Review by loganharbour — DISMISSED** [review]
Mac issues (not on Brandon), also need to time getting in with https://github.com/idaholab/moose/pull/26023

**Comment by brandonlangley** [comment]
> Mac issues (not on Brandon), also need to time getting in with #26023

@loganharbour -

It looks like @milljm fixed the SDK issues for that [Conda build (Intel Mac)](https://civet.inl.gov/job/1884061/) job so everything's green now.

**Comment by loganharbour** [comment]
> > Mac issues (not on Brandon), also need to time getting in with #26023
> 
> 
> 
> @loganharbour -
> 
> 
> 
> It looks like @milljm fixed the SDK issues for that [Conda build (Intel Mac)](https://civet.inl.gov/job/1884061/) job so everything's green now.
> 
> 

That's only part of it unfortunately. Package transfers (conda) aren't currently working. This updates packages, so it's a no go for now

**Comment by loganharbour** [comment]
Once we get a next -> devel merge (hopefully shortly), I'll get this in.

**Comment by brandonlangley** [comment]
> Once we get a next -> devel merge (hopefully shortly), I'll get this in.

@loganharbour - sounds good.

**Comment by loganharbour** [comment]
Ok, we're done here! We had some conda issues come up that broke the new apptainer builds that got triggered by this PR. It looks like anaconda fixed them magically in the background for us.

Will merge once we pass.

**Comment by loganharbour** [comment]
...maybe not

```
outputs/format.hit: Working Directory: /opt/civet0/build/moose/test/tests/outputs/format
outputs/format.hit: Running command: python3 -m unittest -v test_hit_output.TestInputFileFormat
outputs/format.hit: testInputFileFormat (test_hit_output.TestInputFileFormat)
outputs/format.hit: Some basic checks to see if some data ... libc++abi: terminating due to uncaught exception of type hit::Error: cannot convert field 'MeshDivisions/*/<types>/<CylindricalGridDivision>/cylinder_axial_min' value '2.22507e-308' to float
outputs/format.hit: 
outputs/format.hit: 
outputs/format.hit: Exit Code: -6
outputs/format.hit: ################################################################################
outputs/format.hit: Tester failed, reason: CRASH
outputs/format.hit: 
[7.833s]     FAIL outputs/format.hit FAILED (CRASH)
```

in https://civet.inl.gov/job/1895340/

**Comment by loganharbour** [comment]
Going to get this in after https://github.com/idaholab/moose/pull/26102.

There's still one issue that still needs to be resolved

**Comment by brandonlangley** [comment]
@loganharbour @GiudGiud -

This request to review a small update and its reason explanation may seem a bit long winded, but please bear with me.

```
outputs/format.hit: Working Directory: /opt/civet0/build/moose/test/tests/outputs/format
outputs/format.hit: Running command: python3 -m unittest -v test_hit_output.TestInputFileFormat
outputs/format.hit: testInputFileFormat (test_hit_output.TestInputFileFormat)
outputs/format.hit: Some basic checks to see if some data ... libc++abi: terminating due to uncaught exception of type hit::Error: cannot convert field 'MeshDivisions/*/<types>/<CylindricalGridDivision>/cylinder_axial_min' value '2.22507e-308' to float
outputs/format.hit: 
outputs/format.hit: 
outputs/format.hit: Exit Code: -6
outputs/format.hit: ################################################################################
outputs/format.hit: Tester failed, reason: CRASH
outputs/format.hit: 
[7.833s]     FAIL outputs/format.hit FAILED (CRASH)
```

That new failure is from this `cylinder_axial_min` parameter being [added](https://github.com/idaholab/moose/commit/8028c8ba832cc021549d53989fc8f7dc031a7210#diff-67ce8524848b20fb573dcbe0e7240d0b04462a5fd162119d275d303801513276R39-R42) after all the tests originally passed for this PR:

```
params.addRangeCheckedParam<Real>("cylinder_axial_min",
                                  std::numeric_limits<Real>::min(),
                                  "cylinder_axial_min>0",
                                  "Minimum axial coordinate");
```

It stems from it specifying `std::numeric_limits<Real>::min()` as its default value.

That `format.hit` test basically does a `--dump` of the entire syntax and does things like checking for duplicates, etc.

This involves getting the value of each parameter from the hit `--dump` output where defaults are listed as their values.

So this new parameter is listed in the `--dump` output as:

```
[<CylindricalGridDivision>]
  . . .
  cylinder_axial_min = 2.22507e-308
  . . .
[]
```

You can see that test passed fine for the [Conda (Ubuntu)](https://civet.inl.gov/job/1895342/) job but failed for the [Conda (ARM Mac)](https://civet.inl.gov/job/1895340/) job.

MOOSE retrieves its values by type by doing the following (using `double` as an explicit example):

```
  std::string string_value
  std::istringstream iss(str_value);
  double converted_val;
  if ((iss >> converted_val).fail() || !iss.eof())
    throw Error("cannot convert field '" + fullpath() + "' value '" + str_value + "' to float");
  return converted_val;
```

Even though `std::numeric_limits<double>::min()` is `2.22507e-308`, `ss >> double` fails for that value on ARM Mac.

On Linux (and possibly on Intel Mac, though unverified) this `ss >> double` works fine as expected for the `min()` value.

I ran my own set of tests on my M1 ARM Mac to narrow down where the cutoff is that will work for this and found:

```
2.22507e-308 (i.e., std::numeric_limits<double>::min()) - cannot convert to double using ss >> double

1.00000e-308                                            - cannot convert to double using ss >> double

9.99999e-307                                            - converts to double fine using ss >> double
```

I ran these same tests on a Linux box and all three values converted as expected via the `ss >> double` operation.

I would expect the `std::numeric_limits<double>::min()` value to work fine streaming into a double across all systems.

So the cutoff observed above on ARM Mac between `9.99999e-307` and `1.00000e-308` is definitely strange.

- - - - -

The reason the test passes on `devel` is because `kind` says `2.22507e-308` is a string and it's retrieved fine as a string.

However, I used the `devel` branch to test an actual input file that specifies this default value explicitly like:

```
[MeshDivisions]
  [cylindrical_div]
    type = CylindricalGridDivision
    . . .
    cylinder_axial_min = 2.22507e-308
    . . .
  []
[]
```

And running this input that explicitly specifies the default value for the parameter does immediately fail on `devel` with:

```
*** ERROR ***
/path/to/input.i:line.col: invalid float syntax for parameter: MeshDivisions/cylindrical_div/cylinder_axial_min=2.22507e-308
```

There isn't currently anywhere else in MOOSE that uses `std::numeric_limits<Real>::min()` to set a parameter default.

Parameters do use `std::numeric_limits<Real>::lowest()` which is the most negative instead of the one closest to zero.

And these `lowest()` (most-negative) values are handled just fine, but the `min()` (closest-to-zero) do not work on ARM.

I don't know why ARM Mac won't stream `std::numeric_limits<double>::min()` to a double but `9.99999e-307` will fine.

- - - - -

But all of this is to explain why I'm proposing an update like in latest recent [commit](https://github.com/idaholab/moose/pull/26069/commits/59f0c2b64cd91ac4533e75c77a11d8bff8953dd8) here that changes the default to zero.

I didn't see `_min_z` used as a divisor anywhere at a brief glance, but it's possible there are other reasons to not use zero.

@GiudGiud - Since this is your code, could you check this last commit and let me know if this is an acceptable change?

Note that I also changed the non-infinite size check to use:

```
 _max_z == std::numeric_limits<Real>::max()
```

instead of this which I perceived to be possibly be a typo:

```
 _max_z == std::numeric_limits<Real>::min()
```

but it was at a quick look and this change may be incorrect.

- - - - -

@loganharbour - I changed the WASP update commit SHA and will fix `tests.versioner` after hearing from @GiudGiud.

File: framework/src/meshdivisions/CylindricalGridDivision.C
Comment (GiudGiud) [inline framework/src/meshdivisions/CylindricalGridDivision.C:40]: no that wont work.

The cylinder was previously infinite in the minus-axial direction. Now it s not only finite but of 0 height in the minus-axial direction

Comment (GiudGiud) [reply framework/src/meshdivisions/CylindricalGridDivision.C:40]: we do need a big negative number here. It does not have to be -10^300 though. ANything pretty pretty big is good enough

Comment (GiudGiud) [reply framework/src/meshdivisions/CylindricalGridDivision.C:40]: oh and the range check was wrong ... that was my fault!

Comment (GiudGiud) [reply framework/src/meshdivisions/CylindricalGridDivision.C:40]: std::numeric_limits<Real>::lowest() would have been right then

I m afraid I never used min() before amd I guess I got it wrong

Comment (GiudGiud) [reply framework/src/meshdivisions/CylindricalGridDivision.C:40]: -10^30 is probably big (small) enough tbh

Comment (brandonlangley) [reply framework/src/meshdivisions/CylindricalGridDivision.C:40]: @GiudGiud -

Okay great, `std::numeric_limits<Real>::lowest()` actually works fine for `ss >> double` on ARM Mac.

I'll simply change it to that, and also fix the `_max_z == std::numeric_limits<Real>::max()` range check.

The default will now be `-1.79769e+308` (super negative) instead of `2.22507e-308` (positive almost zero)

The only issue now is the current range check function `cylinder_axial_min>0` in:

```
params.addRangeCheckedParam<Real>("cylinder_axial_min",
                                  std::numeric_limits<Real>::min(),
                                  "cylinder_axial_min>0",
                                  "Minimum axial coordinate");
```

will always fail for the default negative value now because that check says it must be positive.

Would you like me to change that range check function to something like `cylinder_axial_min<0` instead.

This would, of course, just say the value has to be negative instead of positive.

Then the only issue is that one test input problem [test/tests/meshdivisions/cylindrical_division.i](https://github.com/idaholab/moose/blob/next/test/tests/meshdivisions/cylindrical_division.i#L29) specifies:

```
[MeshDivisions]
  [cylindrical_div]
    type = CylindricalGridDivision
    . . .
    # Extent of the cylinder
    cylinder_axial_min = 0.5
    cylinder_axial_max = 2
    r_min = 0.5
    r_max = 4
  []
[]
```

And this positive `cylinder_axial_min` test input value then fails that `cylinder_axial_min<0` check with:

```
*** ERROR ***
Range check failed for parameter MeshDivisions/cylindrical_div/cylinder_axial_min
       Expression: cylinder_axial_min<0
       Value: 0.5
```

 - a) Should `cylinder_axial_min` really be a non-range-checked parameter that can be any value (-inf, +inf)?
 
 - b) Or should it be a range-checked negative value and [test/tests/meshdivisions/cylindrical_division.i](https://github.com/idaholab/moose/blob/next/test/tests/meshdivisions/cylindrical_division.i#L29) is bad?

Which of those above choices should be the change here? If (b) is the correct change, then with:

```
cylinder_axial_min = 0.5
```

in that test input, what do you suggest I change it to? Should this be the new value:

```
cylinder_axial_min = -0.5
```

Comment (GiudGiud) [reply framework/src/meshdivisions/CylindricalGridDivision.C:40]: no range check is the way to go here.

and in the constructor we should check min < max that s it

Comment (brandonlangley) [reply framework/src/meshdivisions/CylindricalGridDivision.C:40]: @GiudGiud -

Okay cool thanks, I'll make this update. And that `_min_z < _max_z` error check is already in the constructor.

Last question, I guess this `cylinder_axial_max` should also be changed to be a non-range-checked parameter?

```
  params.addRangeCheckedParam<Real>("cylinder_axial_max",
                                    std::numeric_limits<Real>::max(),
                                    "cylinder_axial_max>0",
                                    "Maximum axial coordinate");
```

And its value can be also be any real number positive or negative (as long as min < max), right?

Comment (GiudGiud) [reply framework/src/meshdivisions/CylindricalGridDivision.C:40]: yes that seems good too.

Comment (brandonlangley) [reply framework/src/meshdivisions/CylindricalGridDivision.C]: @GiudGiud -

My latest [commit](https://github.com/idaholab/moose/pull/26069/commits/968f8c6b6be1eec24a7b64d664d296370753b50f) just made these fixes in regard to those cylindrical axial parameters:

 - fixed the non-infinite size check to use `_max_z == std::numeric_limits<Real>::max()` instead of `min()`
 - updated `cylinder_axial_min` default value to `std::numeric_limits<Real>::lowest()` instead of `min()`
 - updated `cylinder_axial_min` to be a regular type parameter instead of a range-checked type parameter
 - updated `cylinder_axial_max` to be a regular type parameter instead of a range-checked type parameter
 
 Could you double-check that commit to make sure all looks good there?
 
- - - - -

@loganharbour -

I fixed the SHA in `versioner_hashes.yaml` too, since I rebased on devel which changed the WASP commit.

That should fix the `tests.versioner` failure, so this is good to again from my side if @GiudGiud approves.

**Review by loganharbour — DISMISSED** [review]
Requires https://github.com/idaholab/moose/pull/26102 (going in today) + removal of the wasp update.

WASP update can remain for now so that we can complete testing. Once #26102 is in, I'll rebase to remove the update and just have the optimization and `CylindricalGridDivision` fix.

**Review by GiudGiud — APPROVED** [review]
division changes are great

**Review by GiudGiud — APPROVED** [review]

**Review by loganharbour — APPROVED** [review]

**Pull request merged**
